trie.h

```diff
@@ -0,0 +1,37 @@
+#ifndef TRIE_H
+#define TRIE_H
+
+#include <map>
+#include <vector>
+#include <string>
+#include <memory>
+
+using std::string;
+using std::map;
+using std::vector;
+using std::shared_ptr;
+
+class Trie {
+private:
+    struct Node {
+        map<char, shared_ptr<Node>> children;
```

Comment:
I had to look up online what a trie is supposed to look like in memory. It'd be helpful to have a simple example that explains what Node and Node::children are for.